lib/compat/wordpress-6.1/class-gutenberg-rest-edit-site-export-controller.php

```diff
@@ -0,0 +1,189 @@
+			}
+			if ( ! wp_is_writable( $fileToUpdate ) ) {
+				throw new Exception("Theme file is not writeable: " . $fileToUpdate);
+			}
+			file_put_contents(
+				$fileToUpdate,
+				$template->content
+			);
+		}
+
+		foreach ( $user_customisations['template_parts'] as $template_part ) {
+			$fileToUpdate = get_template_directory() . '/parts/' .
+				$template_part->slug . '.html';
+			if ( ! file_exists( $fileToUpdate ) ) {
+				throw new Exception("Theme file does not exist: " . $fileToUpdate);
+			}
+			if ( ! wp_is_writable( $fileToUpdate ) ) {
+				throw new Exception("Theme file is not writeable: " . $fileToUpdate);
+			}
+			file_put_contents(
+				$fileToUpdate,
+				$template_part->content
+			);
+		}
+	}
+
+	protected function update_theme_json_file() {
+		$fileToUpdate = get_template_directory() . '/theme.json';
+		if ( ! file_exists( $fileToUpdate ) ) {
+			throw new Exception("Theme file does not exist: " . $fileToUpdate);
+		}
+		if ( ! wp_is_writable( $fileToUpdate ) ) {
+			throw new Exception("Theme file is not writeable: " . $fileToUpdate);
+		}
+		file_put_contents(
+			$fileToUpdate,
+			gutenberg_export_theme_json()
+		);
+	}
+
+	protected function get_user_customisations() {
+
+		$templates = gutenberg_get_block_templates();
+		$template_parts = gutenberg_get_block_templates ( array(), 'wp_template_part' );
+
+		$exported_templates = [];
+		$exported_template_parts = [];
+
+		foreach ( $templates as $template ) {
+			if ( $template->source !== 'custom' ) {
+				continue;
+			}
+			$template = $this->clean_template( $template );
+			$exported_templates[] = $template;
+		}
+
+		foreach ( $template_parts as $template_part ) {
+			if ( $template_part->source !== 'custom' ) {
+				continue;
+			}
+			$template_part = $this->clean_template( $template_part );
+			$exported_template_parts[] = $template_part;
+		}
+
+		return array(
+			'templates'=>$exported_templates,
+			'template_parts'=>$exported_template_parts
+		);
+
+	}
+
+	/*
+	 * Clean a template or template part by reversing unicode escaping
+	 * and removing the theme attribute from the content
+	 */
+	function clean_template( $template ) {
+		$template->content = _remove_theme_attribute_in_block_template_content(
+			$template->content
+		);
+
+		// `serialize_block_attributes` has "unicode escape sequence
+ 		// substitution for characters which might otherwise interfere with embedding
+ 		// the result in an HTML comment".
+		// This reverses that escaping for saving back to files.
+		// @see `wp-includes/blocks.php`
+		$template->content = str_replace( '\\u002d\\u002d', '--', $template->content );
+		$template->content = str_replace( '\\u003c', '<', $template->content );
+		$template->content = str_replace( '\\u003e', '>', $template->content );
+		$template->content = str_replace( '\\u0026', '&', $template->content );
+		$template->content = str_replace( '\\u0022', '"', $template->content );
+
+		return $template;
+	}
+
+	function clear_user_customizations() {
```

Comment:
I wonder if we already have something that does this.

Reply:
Can we use this? https://github.com/WordPress/gutenberg/blob/6bcacdd811b4755e8f9daa87d0414af87e732358/packages/edit-site/src/components/global-styles/hooks.js#L25

Reply:
Not right now as it returns a buggy `canReset` which is always true. You can see that the reset menu item in the global styles panel is always active. But I'll see.